scio-core/src/main/scala-3/com/spotify/scio/schemas/To.scala

```diff
@@ -18,15 +18,92 @@
 package com.spotify.scio.schemas
 
 import org.apache.beam.sdk.schemas.{SchemaCoder, Schema => BSchema}
+import BSchema.{ FieldType => BFieldType }
 
 import scala.compiletime._
 import scala.deriving._
 import scala.quoted._
+import scala.reflect.ClassTag
 
 object ToMacro {
-  def safeImpl[I, O](si: Expr[Schema[I]])(implicit q: Quotes): Expr[To[I, O]] = {
-    ???
+
+  def safeImpl[I: scala.quoted.Type, O: scala.quoted.Type](
+    iSchema: Expr[Schema[I]],
+    oSchema: Expr[Schema[O]]
+  )(using Quotes): Expr[To[I, O]] = {
+    import scala.quoted.quotes.reflect.{report, TypeRepr}
+
+    (interpret[I] , interpret[O]) match {
+      case (None, None) => report.throwError(
+        s"""
+        |Could not interpret input schema:
+        |  ${iSchema.show}
+        |Could not interpret output schema:
+        |  ${oSchema.show}
+        |""".stripMargin
+      )
+      case (None, _) => report.throwError("Could not interpret input schema: " + iSchema.show)
+      case (_, None) => report.throwError("Could not interpret output schema: " + oSchema.show)
+      case (Some(sIn), Some(sOut)) =>
+        val schemaOut: BSchema = SchemaMaterializer.fieldType(sOut).getRowSchema()
+        val schemaIn: BSchema = SchemaMaterializer.fieldType(sIn).getRowSchema()
+        val classTagOpt = Expr.summon[ClassTag[O]]
+        if (classTagOpt.isEmpty) {
+          report.throwError(s"Could not summon Expr[ClassTag[${TypeRepr.of[O].show}]]")
+        }
+        val classTag = classTagOpt.get
+        To.checkCompatibility(schemaIn, schemaOut)('{ To.unchecked[I, O](using $iSchema, $oSchema, $classTag) })
+          .fold(message => report.throwError(message), identity)
+    }
   }
+
+  private def sequence[T](ls: List[Option[T]]): Option[List[T]] =
+    if ls.exists(_.isEmpty) then None
+    else Some(ls.collect { case Some(x) => x })
+
+  private def interpret[T: scala.quoted.Type](using Quotes): Option[Schema[T]] = 
+    Type.of[T] match {
```

Comment:
Unless I'm missing something this means that the possible set of supported schema types is effectively limited to those already defined in scio. 

If a user adds a `given` instance of `Schema` for type that isn't supported (say for example a Java class), the derivation will simply ignore it. This could be a problem for aliases support in Schema but probably something we can live with. We don't really expect users to define their own instances of `Schema`.

I'd be curious to see how the implementation looks like for Java beans and Avro's `SpecificRecord`.

Reply:
> Unless I'm missing something this means that the possible set of supported schema types is effectively limited to those already defined in scio.

Yes, that's the assumption here.

> If a user adds a given instance of Schema for type that isn't supported (say for example a Java class), the derivation will simply ignore it. This could be a problem for aliaswa support in Schema but probably something we can live with. We don't really expect users to define their own instances of Schema.

It's still possible to support those use cases by registering custom interpreters to be tried as a last resort. Indeed, it would complicate things and can be subtle: think the case where the user defines the schema and use it in the same project.

> This could be a problem for aliaswa support in Schema

It's not clear what you mean above. Could you please clarify?

> I'd be curious to see how the implementation looks like for Java beans and Avro's SpecificRecord.

@vincenzobaz and @MaximeKjaer are going to work on it. I'll let them report progress on it.

Reply:
Hey. Yeah sorry I wanted to talk about [LogicalType](https://beam.apache.org/releases/javadoc/2.20.0/org/apache/beam/sdk/schemas/Schema.LogicalType.html) actually but got the exact term used in beam wrong... I think that's the only case for which a user might want to define a custom type. Last time I checked the implementation was broken in beam and fixing it did not seem to be the priority so I guess we can live without it.

The most common use case for us would be to convert from avro's `SpecificRecord` to case classes and vice-versa so this one needs to be working really well.

Reply:
I was checking out how we could handle java beans and I hit an important blocker I think.
It is really easy to work around the implicit instance of `IsJavaBean[T]` because an instance of such type it is just an empty evidence, moreover the code to instantiate it is already ported to macro. It is straightforward to add a case to the pattern match which could like this:

```scala
 case _ if TypeRepr.of[T].typeSymbol.flags.is(Flags.JavaDefined) && Try(IsJavaBean.isJavaBeanImpl[T]).isSuccess => 
    ???
 ```
I also remarked that we only need a `BSchema` for the compatibility check, so we might avoid altogether reaching for a `RawRecord[T]`.

The major pain point is that `JavaBeanSchema.schemaFor` which is used to create the `BSchema` in question requires a `Class[T]` (which explains the implicit requirement of a `ClassTag[T]`). `Class[T]` is available only at runtime, so it is not possible to have instances of it in the macro.

I am not sure how we can handle this problem besides maybe implementing our own schema derivation for java beans.
What do you think?

Reply:
This blocker might affect `AvroInstances.avroSchema` as well given that it relies on an implicit `ClassTag`

Reply:
@jto I tested the Scala2 version of the macro with JavaBean to inspect the behavior. You can see the result of the test [here](https://github.com/scalacenter/scio/runs/2438868791?check_suite_focus=true#step:5:314).
From this I conclude that we cannot use Java's runtime reflection api (`Class<T>` therefore `ClassTag[T]`) in our compile time macro. The same conclusion is likely to be true for Avro's Schema derivation which relies on the same logic.

It could be possible to proceed similarly to case classes: implement our own logic which iterates over the fields of the bean (actually the getters) and recursively interpret their type. However this might result in a compile time `Schema` which diverges from the one generated at runtime by the `beam` API, over which we have no control.
Because of this risk and of the potential corner cases hit by `interpret` call, I think you should investigate the usage of java bean among scio users to decide whether such a time investment would be worthwhile.
In the meanwhile, it could be useful to abort the compilation (by launching an error in the macro) when a macro is asked to check the compatibility of two schemas at compiletime as the error could be puzzling to a library user.

What do you think?

Reply:
@jto The test seems to suggest that `To.safe[S, T]` in Scala 2 never worked for Java Beans if the type `S` or `T` is defined in the same project as to where `To.safe[S, T]` is used.

Reply:
Indeed. The Java class needs to have already been compiled in a previous compilation unit.

Reply:
> However this might result in a compile time Schema which diverges from the one generated at runtime by the beam
 This might be the case but seems unlikely. I think r-eimplementing it is an acceptable solution. We can always have tests to validate the the Schema is the same as what beam would derive.

Reply:
IIRC for `Avro` we actually only need to access the avro Schema. There's a schema field in the generated code so it should be possible to access it at compile time since it's just a String ?

Reply:
For avro I see two methods:

- `implicit def avroSchema[T <: SpecificRecord: ClassTag]: Schema[T]` which relies on the `ClassTag` to feed a `Class` to `TypeDescriptor` and therefore suffers from the same problem as the Java bean discussed here.
- `def fromAvroSchema(schema: org.apache.avro.Schema): Schema[GenericRecord]` which is not generic nor implicit.

I imagine that you refer to the second one. In this case we can pattern match on the call of `fromAvroSchema` but we have not compile time information about `schema: avro.Schema`, so I am not sure about how to obtain the `Schema[GenericRecord]`

Reply:
@jto commit 93de09c proposes an implementation of compile time schema derivation for java beans similar to the one used for case classes. It seems to be accepted by (simple) tests.
Does it look reasonable to you?

Reply:
Yep. Seems reasonable :)